zkevm-circuits/src/state_circuit/lexicographic_ordering.rs

```diff
@@ -0,0 +1,316 @@
+            ]
+        });
+        assert!(meta.degree() <= 16);
+        meta.create_gate("diff_1 is zero iff all 15 possible values are 0", |meta| {
+            let selector = meta.query_fixed(selector, Rotation::cur());
+            let cur = Queries::new(meta, &config, Rotation::cur());
+            let prev = Queries::new(meta, &config, Rotation::prev());
+            vec![
+                // all 15 possible values are 0 iff the final linear combination is 0
+                selector * diff_1_is_zero.clone() * diff_1_possible_values(cur, prev)[14].clone(),
+            ]
+        });
+        meta.create_gate("diff_2 is one of 15 values", |meta| {
+            let selector = meta.query_fixed(selector, Rotation::cur());
+            let cur = Queries::new(meta, &config, Rotation::cur());
+            let prev = Queries::new(meta, &config, Rotation::prev());
+            let diff_2 = meta.query_advice(diff_2, Rotation::cur());
+            vec![
+                selector
+                    * diff_2_possible_values(cur, prev)
+                        .iter()
+                        .map(|e| diff_2.clone() - e.clone())
+                        .fold(1.expr(), Expression::mul),
+            ]
+        });
+        assert!(meta.degree() <= 16);
+        meta.lookup_any("diff_1 fits into u16", |meta| {
+            let diff_1 = meta.query_advice(diff_1, Rotation::cur());
+            vec![(diff_1, meta.query_fixed(u16_range, Rotation::cur()))]
+        });
+        meta.lookup_any("diff_1 is zero or diff_2 fits into u16", |meta| {
+            let diff_2 = meta.query_advice(diff_2, Rotation::cur());
+            vec![(
+                diff_1_is_zero * diff_2,
+                meta.query_fixed(u16_range, Rotation::cur()),
+            )]
+        });
+        assert!(meta.degree() <= 16);
+        meta.create_gate("diff_2 is not zero", |meta| {
+            let selector = meta.query_fixed(selector, Rotation::cur());
+            vec![(selector * diff_2_is_zero)]
+        });
+        assert!(meta.degree() <= 16);
+
+        config
+    }
+
+    pub fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        cur: &Rw,
+        prev: &Rw,
+    ) -> Result<(), Error> {
+        // this doesn't make sense that we have to "construct" the chip every time we
+        // assign?
+        let diff_1_is_zero_chip = IsZeroChip::construct(self.config.diff_1_is_zero.clone());
+        let diff_2_is_zero_chip = IsZeroChip::construct(self.config.diff_2_is_zero.clone());
+
+        let cur_be_limbs = rw_to_be_limbs(cur);
+        let prev_be_limbs = rw_to_be_limbs(prev);
+
+        let find_result = cur_be_limbs
+            .iter()
+            .zip(&prev_be_limbs)
+            .enumerate()
+            .find(|(_, (a, b))| a != b);
+        let (index, (cur_limb, prev_limb)) = find_result.expect("repeated rw counter");
+
+        let mut diff_1 = F::from((cur_limb - prev_limb).into());
```

Comment:
Doesn't this `diff_1` computation imply `diff_1` can only be `Ai - Bi` and not one of the following (cumulative) values specified above in the comment (and also used in the constraint):
 // C0 = A0 - B0,
 // C1 = C0 << 16 + A1 - B1,
 // ...
 // C14 = C13 << 16 + A14 - B14,

The constraints don't fail because Ci are 0 by having `diff_1` computed this way.

The `diff_1` computation implies only one change can occur in the first 15 limbs. Could be some explanation added why this is so? Also, specs repo doesn't seem to have info how lexicographic ordering is to be implemented, could that be added?

Reply:
I think it's more clear to view it the other way around. The only way that `diff_1` can only be one of the 15 values `C0, ..., C14`, and that `diff_1` can fit into 16 bits is for `diff_1` to be the first `Ai - Bi` such that `Aj = Bj` for j < i. 

You're right that multiple difference can occur between the first 15 pairs of limbs, but (I think) that the constraints are such that `diff_1` has to be the value of the most significant non-zero limb difference.

The specs are for lexicographic ordering in the spec repo are in the code here: https://github.com/scroll-tech/zkevm-specs/blob/master/src/zkevm_specs/state.py#L355. I plan to update that section and the book once the implementation is finalized here. I had to make some changes to how the bit-packing of the keys worked so that everything could fit into 2 field elements.

Reply:
I think I got it now, thanks!

I think it would be helpful to add such details as above to the comments and docs. Perhaps write something like (if I understood correctly):

Assignment: `diff_i` is change in the 15 limbs that occurs first (if `prev = (2, 1, 2, ...), curr = (2, 1, 6, ...)`, then `diff_i = 4`.
Constraint: `diff_i` will be equal to the first non-zero `Ci`.

Perhaps you could rename `diff_1`, `diff_2` to `first_diff_high`, `first_diff_low`?

Reply:
I've renamed them to `upper_limb_difference` and `lower_limb_difference`.

Reply:
Cool, thanks!